pkg/deploy/controller/generictrigger/factory.go

```diff
@@ -4,67 +4,187 @@ import (
+	c := &DeploymentTriggerController{
+		dn: oc,
+		rn: kc,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		codec: codec,
+	}
+
+	c.dcStore.Indexer = dcInformer.GetIndexer()
+	dcInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addDeploymentConfig,
+		UpdateFunc: c.updateDeploymentConfig,
+	})
+
+	streamInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStream,
+		UpdateFunc: c.updateImageStream,
+	})
+
+	c.dcStoreSynced = dcInformer.HasSynced
+	c.streamStoreSynced = streamInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *DeploymentTriggerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the image stream and dc stores to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down deployment trigger controller")
+	c.queue.ShutDown()
 }
 
-// Create creates a DeploymentTriggerController.
-func (factory *DeploymentTriggerControllerFactory) Create() controller.RunnableController {
-	deploymentConfigLW := &cache.ListWatch{
-		ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).List(options)
-		},
-		WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).Watch(options)
-		},
-	}
-	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 2*time.Minute).Run()
-
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartRecordingToSink(factory.KubeClient.Events(""))
-
-	triggerController := NewDeploymentTriggerController(factory.Client, factory.KubeClient, factory.Codec)
-
-	return &controller.RetryController{
-		Queue: queue,
-		RetryManager: controller.NewQueueRetryManager(
-			queue,
-			cache.MetaNamespaceKeyFunc,
-			func(obj interface{}, err error, retries controller.Retry) bool {
-				utilruntime.HandleError(err)
-				if _, isFatal := err.(fatalError); isFatal {
-					return false
-				}
-				if retries.Count > 0 {
-					return false
-				}
-				return true
-			},
-			flowcontrol.NewTokenBucketRateLimiter(1, 10),
-		),
-		Handle: func(obj interface{}) error {
-			config := obj.(*deployapi.DeploymentConfig)
-			return triggerController.Handle(config)
-		},
+func (c *DeploymentTriggerController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.dcStoreSynced() || !c.streamStoreSynced() {
```

Comment:
I see why you have to wait for the DCs to sync, but why do you have to wait for the streamStore to sync?